app/javascript/components/visualization/ScatterPlot.js

```diff
@@ -274,11 +278,36 @@ function getPlotlyTraces({
       styles: legendEntries
     }]
   } else {
+    // for non-clustered plots, we pass in a single trace with all the points
+    let colors
+    if (isGeneExpressionForColor) {
+      // sort the points by order of expression
+      const expressionsWithIndices = data.expression.map((val, i) => [val, i])
+      expressionsWithIndices.sort((a, b) => a[0] - b[0])
+      // now that we know the indices, reorder the other data arrays
+      trace.x = expressionsWithIndices.map(ei => data.x[ei[1]])
```

Comment:
I'm with @eweitz on this - I think calling `map` repeatedly will not scale well for large plots.  Can we reduce this to a single `map` call, and update all traces at the same time?  For example:
```suggestion
      expressionsWithIndices.map(ei => {
        trace.x.concat(data.x[ei[1]])
        trace.y.concat(data.y[ei[1]])
        ...
      })
```

Reply:
Implementing this on the server would be fastest by a dramatic margin.  That way, we could sort _once per file info update_, rather than (at best) once per page load.  Cache invalidation could presumably just reuse what we already have in place for this endpoint.

Reply:
I'm not sure we can do that, since all of the visualization data will depend on the cluster/annotation/gene requested.  We'd have to pre-cache everything, which will never scale.  If we wanted to do this server side at runtime, I'd had the idea previously that we could first plot all of the 0 values, and then plot the significant expression.  Since we only store the significant expression values in Mongo, we can infer all the 0 values from a quick set difference of `(cells in current plot - expressed cells for genes)`, put those in the traces _first_, and then the actual expression values.  It's not quite sorted by expression, but it does put all of the non-zero values last.

Reply:
Good point.  Sorting at ingest seems fastest, in that case, but your proposal is likely better effort/value.

Reply:
Interesting, my thought was that separate maps are better since then the array allocations (which I suspect are by far the most time-intensive part of this operation) can happen all at once, rather than via a series of pushes. But you're right that investigation is warranted